src/core/algorithms/order/order.cpp

```diff
@@ -0,0 +1,366 @@
+                if (valid_.find(lhs) == valid_.end()) {
+                    valid_[lhs] = {};
+                }
+                valid_[lhs].insert(rhs);
+                bool lhs_unique = typed_relation_->GetNumRows() ==
+                                  sorted_partitions_[lhs].sorted_partition.size();
+                if (lhs_unique) {
+                    candidate_sets_[lhs].erase(rhs);
+                }
+            } else if (candidate_validity == +ValidityType::swap) {
+                candidate_sets_[lhs].erase(rhs);
+            } else if (candidate_validity == +ValidityType::merge) {
+                if (merge_invalidated_.find(lhs) == merge_invalidated_.end()) {
+                    merge_invalidated_[lhs] = {};
+                }
+                merge_invalidated_[lhs].insert(rhs);
+            }
+        }
+    }
+    MergePrune();
+}
+
+std::vector<AttributeList> Order::Extend(AttributeList const& lhs, AttributeList const& rhs) const {
+    std::vector<AttributeList> extended_rhss;
+    for (auto const& single_attribute : single_attributes_) {
+        if (AreDisjoint(single_attribute, lhs) && AreDisjoint(single_attribute, rhs)) {
+            AttributeList extended(rhs);
+            extended.push_back(single_attribute[0]);
+            extended_rhss.push_back(extended);
+        }
+    }
+    return extended_rhss;
+}
+
+bool Order::IsMinimal(AttributeList const& a) const {
+    for (auto const& [lhs, rhs_list] : valid_) {
+        for (AttributeList const& rhs : rhs_list) {
+            auto it_rhs = std::search(a.begin(), a.end(), rhs.begin(), rhs.end());
+            if (it_rhs == a.end()) {
+                continue;
+            }
+            if (std::search(it_rhs + rhs.size(), a.end(), lhs.begin(), lhs.end()) != a.end()) {
+                return false;
+            }
+            auto it_lhs = std::search(a.begin(), it_rhs, lhs.begin(), lhs.end());
+            if (it_lhs + lhs.size() == it_rhs) {
+                return false;
+            }
+        }
+    }
+    return true;
+}
+
+void Order::UpdateCandidateSets() {
+    unsigned int level_num = lattice_->GetLevelNumber();
+    if (level_num < 3) {
+        return;
+    }
+    CandidateSets next_candidates;
+    for (auto const& [lhs, rhs_list] : candidate_sets_) {
+        next_candidates[lhs] = {};
+        if (lhs.size() != level_num - 1) {
+            for (AttributeList const& rhs : rhs_list) {
+                if (InUnorderedMap(valid_, lhs, rhs)) {
+                    continue;
+                }
+                std::vector<AttributeList> extended_rhss = Extend(lhs, rhs);
+                for (AttributeList const& extended : extended_rhss) {
+                    if (lhs.size() > 1) {
+                        AttributeList lhs_max_prefix = MaxPrefix(lhs);
+                        std::vector<AttributeList> extended_prefixes = GetPrefixes(extended);
+                        auto prefix_is_valid = [&](AttributeList const& extended_prefix) {
```

Comment:
[Don't use default capture by reference for such a big scope](https://google.github.io/styleguide/cppguide.html#Lambda_expressions), it'll lead to hard-to-catch bugs with dangling references and is simply not needed here. Capturing explicitly `this` and `lhs_max_prefix` is enough